sdk/clientcore/core/src/main/java/io/clientcore/core/http/models/PagedIterable.java

```diff
@@ -0,0 +1,206 @@
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * This class provides utility to iterate over {@link PagedResponse} using {@link Stream} and {@link Iterable}
+ * interfaces.
+ *
+ * @param <T> The type of items in the page.
+ */
+public final class PagedIterable<T> implements Iterable<T> {
+
+    private final Function<String, PagedResponse<T>> pageRetriever;
+
+    /**
+     * Creates an instance of {@link PagedIterable} that consists of only a single page. This constructor takes a {@code
+     * Supplier} that return the single page of {@code T}.
+     *
+     * @param firstPageRetriever Supplier that retrieves the first page.
+     */
+    public PagedIterable(Supplier<PagedResponse<T>> firstPageRetriever) {
+        this(firstPageRetriever, null);
+    }
+
+    /**
+     * Creates an instance of {@link PagedIterable}. The constructor takes a {@code Supplier} and {@code Function}. The
+     * {@code Supplier} returns the first page of {@code T}, the {@code Function} retrieves subsequent pages of {@code
+     * T}.
+     *
+     * @param firstPageRetriever Supplier that retrieves the first page.
+     * @param nextPageRetriever Function that retrieves the next page given a continuation token
+     */
+    public PagedIterable(Supplier<PagedResponse<T>> firstPageRetriever,
+        Function<String, PagedResponse<T>> nextPageRetriever) {
+        this.pageRetriever = (continuationToken) -> (continuationToken == null)
+            ? firstPageRetriever.get()
+            : nextPageRetriever.apply(continuationToken);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Iterator<T> iterator() {
+        return iterableByItemInternal().iterator();
+    }
+
+    /**
+     * Retrieve the {@link Iterable}, one page at a time. It will provide same {@link Iterable} of T values from
+     * starting if called multiple times.
+     *
+     * @return {@link Iterable} of a pages
+     */
+    public Iterable<PagedResponse<T>> iterableByPage() {
+        return iterableByPageInternal();
+    }
+
+    /**
+     * Utility function to provide {@link Stream} of value {@code T}.
+     *
+     * @return {@link Stream} of value {@code T}.
+     */
+    public Stream<T> stream() {
+        return StreamSupport.stream(iterableByItemInternal().spliterator(), false);
+    }
+
+    /**
+     * Retrieve the {@link Stream}, one page at a time. It will provide same {@link Stream} of T values from starting if
+     * called multiple times.
+     *
+     * @return {@link Stream} of a pages
+     */
+    public Stream<PagedResponse<T>> streamByPage() {
+        return StreamSupport.stream(iterableByPage().spliterator(), false);
+    }
+
+    private Iterable<T> iterableByItemInternal() {
+        return () -> new PagedByIterator<>(pageRetriever) {
+
+            private final Queue<Iterator<T>> pages = new ConcurrentLinkedQueue<>();
```

Comment:
Do we expect to have more than one page at a time in this queue?

Reply:
I think it should only have at most 1 page.

Maybe this is also for concurrency protection.

Reply:
I've removed all concurrency related syntax/class. And here this is replaced with a single `nextPage` be either `null` of one page.